blindsign/blindrsa/blindrsa.go

```diff
@@ -0,0 +1,213 @@
+package blindrsa
+
+// This package implements the blind RSA protocol based on the CFRG specification:
+// https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-rsa-blind-signatures-02
+
+import (
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/subtle"
+	"errors"
+	"hash"
+	"io"
+	"math/big"
+
+	"github.com/cloudflare/circl/blindsign"
+)
+
+// An RSAVerifier represents a Verifier in the RSA blind signature protocol.
+// It carries state needed to produce and validate an RSA blind signature.
+type RSAVerifier struct {
+	// Public key of the Signer
+	pk *rsa.PublicKey
+
+	// Hash function used in producing the message signature
+	hash hash.Hash
+}
+
+// NewRSAVerifier creates a new RSAVerifier using the corresponding Signer parameters.
+func NewRSAVerifier(pk *rsa.PublicKey, hash hash.Hash) RSAVerifier {
+	return RSAVerifier{
+		pk:   pk,
+		hash: hash,
+	}
+}
+
+func encodeMessageEMSAPSS(message []byte, key *rsa.PublicKey, hash hash.Hash, salt []byte) ([]byte, error) {
+	hash.Write(message)
+	digest := hash.Sum(nil)
+	hash.Reset()
+	emBits := key.N.BitLen() - 1
+	encodedMsg, err := emsaPSSEncode(digest[:], emBits, salt, hash)
+	return encodedMsg, err
+}
+
+func blind(random io.Reader, key *rsa.PublicKey) (*big.Int, *big.Int, error) {
+	randReader := random
+	if randReader == nil {
+		randReader = rand.Reader
+	}
+	r, err := rand.Int(randReader, key.N)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if r.Sign() == 0 {
+		r = bigOne
+	}
+	rInv := new(big.Int).ModInverse(r, key.N)
+	if rInv == nil {
+		return nil, nil, ErrInvalidBlind
+	}
+
+	return r, rInv, nil
+}
+
+func (v RSAVerifier) fixedBlind(message, salt []byte, r, rInv *big.Int) ([]byte, blindsign.VerifierState, error) {
+	encodedMsg, err := encodeMessageEMSAPSS(message, v.pk, v.hash, salt)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	m := new(big.Int).SetBytes(encodedMsg)
+
+	bigE := big.NewInt(int64(v.pk.E))
+	x := new(big.Int).Exp(r, bigE, v.pk.N)
+	z := new(big.Int).Set(m)
+	z.Mul(z, x)
+	z.Mod(z, v.pk.N)
+
+	blindedMsg := z.Bytes()
+
+	return blindedMsg, RSAVerifierState{
+		encodedMsg: encodedMsg,
+		verifier:   v,
+		rInv:       rInv,
+	}, nil
+}
+
+// Blind initializes the blind RSA protocol using an input message and source of randomness. If
+// randomness is not provided, the output signature is deterministic. Otherwise, the signature
+// includes a randomly generated PSS salt whose length equals the size of the underlying hash
+// function.
+//
+// See the specification for more details:
+// https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-rsa-blind-signatures-02#section-5.1.1
+func (v RSAVerifier) Blind(random io.Reader, message []byte) ([]byte, blindsign.VerifierState, error) {
+	var salt []byte
+	if random == nil {
+		salt = nil
+	} else {
+		salt = make([]byte, v.hash.Size())
+		_, err := random.Read(salt)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	r, rInv, err := blind(random, v.pk)
```

Comment:
If `random == nil` then the blinding factor will be chosen deterministically (and thus predictable to the signer). This needs to be called out in the method's comment, as it would seem to violate the security property the API is supposed to provide.

It's one thing for the message encoding to be deterministic, it's another for the blinding factor to be deterministic.